### PR TITLE
octave: bring back java

### PR DESCRIPTION
**Pull request by schoeps**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

Changes
* added java as optional dependency
* brought back proper checking of the build
* require `sundials` (necessary for development and future stable branch)
* require `fig2dev` (this might be arguable due to #16193)
* added dependency on curl (or is this standard on all platforms?)

Question:
* what about Linux people? The new way of ignoring accelerate alternatives is incompatible or not?

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
> brought back proper checking of the build

Can you elaborate on "proper"?

> require sundials (necessary for development and future stable branch)

Let's add it only when it's needed for our stable release, please.

> require fig2dev (this might be arguable due to #16193)

Yes, we don't want an X11 dependency here.





> added dependency on curl (or is this standard on all platforms?)

It's standard on macOS which is the only supported platform here.



> what about Linux people? The new way of ignoring accelerate alternatives is incompatible or not?

That's up to Linuxbrew to decide. We don't support Linux here.

File: Formula/octave.rb [hunk 1]
```diff
@@ -49,7 +49,7 @@ class Octave < Formula
   depends_on "qhull"
   depends_on "qrupdate"
   depends_on "readline"
-  depends_on "sundials" # new dependency for octave > 4.2.1
+  depends_on "sundials" if build.head?
```
Comment (MikeMcQuaid) [inline Formula/octave.rb]: Put in a `head do` block.

**Comment by schoeps** [comment]
> brought back proper checking of the build
>> Can you elaborate on "proper"?

there is a `make check` that is now used again. It runs the test suite and creates a report. The report shows that our binary has many non-expected errors. One should understand why, but at least we should ship the report?
 
> require sundials (necessary for development and future stable branch)
>> Let's add it only when it's needed for our stable release, please.

ok, changed to `if build.head?`

> require fig2dev (this might be arguable due to #16193)
>> Yes, we don't want an X11 dependency here.

as explained in #16193, exporting plots in different formats is core functionality.

**Comment by MikeMcQuaid** [comment]
> as explained in #16193, exporting plots in different formats is core functionality.

Regardless, if Octave can me made to work without X11 that's more important for us as it's not shipped with macOS.

**Comment by ilovezfs** [comment]
Is Java actually needed at build-time to enable the Java option?

**Comment by ilovezfs** [comment]
For fig2dev, [this patch](https://gist.github.com/ilovezfs/8d634d9d717d7c30ee6a31dee6a0cedc) gets it building without x11. @schoeps can you confirm whether a build of fig2dev without x11 works properly for octave's purposes?

**Comment by schoeps** [comment]
> Is Java actually needed at build-time to enable the Java option?

I am afraid yes.

> For fig2dev, this patch gets it building without x11. @schoeps can you confirm whether a build of fig2dev without x11 works properly for octave's purposes?

Guess it should work. I will start rebuilding asap.

**Comment by schoeps** [comment]
As a next step I would really like to brainstorm with you guys on how to get a proper `:blas` treatment... even at the cost that choosing something other than Apple will not allow to use bottles.

**Comment by fxcoudert** [comment]
@schoeps thanks for confirmation that fig2dev without X11 is sufficient for Octave. PR open at https://github.com/Homebrew/homebrew-core/pull/16825

**Comment by fxcoudert** [comment]
sundials is not current in core (but I'm bringing it over: https://github.com/Homebrew/homebrew-core/pull/16832)

**Comment by ilovezfs** [comment]
How does this sound?
```
depends_on :java => ["1.6+", :build, :recommended]
```
so that the Java bits will make it into the bottles without requiring everyone to install Java.

**Comment by schoeps** [comment]
@ilovezfs Sure. I am just surprised... you guys were so reluctant with respect to x11 that your willingness to embrace java is unexpected :)

**Comment by schoeps** [comment]
@ilovezfs why did you remove the checks? I think that this is important for understanding how good the binary actually is. There are still quite some failures that need to be understood.... it should complete with none.

**Comment by ilovezfs** [comment]
@schoeps in homebrew/core we only run `make check` and `make test` for security software such as openssl and gnupg.

**Comment by schoeps** [comment]
@ilovezfs so installing a computational tool that does erronomeus calculation is fine as long as it does not crash... this is just another proof that Octave (or I) do not belong in homebrew/core *sigh*

**Comment by ilovezfs** [comment]
@schoeps take deep breaths.

**Comment by fxcoudert** [comment]
@schoeps we distribute and package software “as is”. Hopefully when they release software upstream performs quality checks. We do not have the manpower to analyse bugs, regressions, and issues for 4000+ software packages. Experience shows that runtime checks within our formulas leads much more often to waste of maintainer time (known bugs, spurious issues, ill-designed tests, etc.) than proper quality control.

**Comment by schoeps** [comment]
@fxcoudert these tests mostly fail because something goes wrong during the packaging (incompatibilities of libraries etc). If tests fail because something is broken than somebody must report it upstream. That is why I think that we need to have an eye on those tests. Octave is not an audioplayer, it‘s a complex calculator. It might even return false values.
Some time ago, there was a bug in `octave.rb` using `veclibfort` that I discovered because some checks with complex numbers failed. That is why I added the additional test. Isn’t that our job as packagers? 

> we do not have the manpower
 
well, I tried to help out supporting a few packages, admintingly with the help of @dpo, @ilovezfs and others.

@ilovezfs sorry, if I was unkind. I am just upset how thing are going with the transition from science to core but instead of annoying you guys, I should probably indeed just set up my own tap as you suggested to @dpo before.

**Comment by ilovezfs** [comment]
>Some time ago, there was a bug in octave.rb using veclibfort that I discovered because some checks with complex numbers failed.

This is something that should be discovered by upstream's CI, not Homebrew's.

**Comment by iMichka** [comment]
> This is something that should be discovered by upstream's CI, not Homebrew's.

+1 on that. Upstream is responsible to run the test suit on different architecture and setups. There are a lot of free CI solutions out there, so there is no excuse for not having it. There are even some teams that even re-run the brew formula installation with a test in their CI.

BTW, the user is also responsible to write his own tests to check the calculation he does with octave are right; they should not blindly believe the results from a tool/library.

I am quite unhappy of all the checks in homebrew-science that sometimes add more work for us, and we even end up disabling some test, after having reported the errors to an unresponsive upstream ... there are better ways to loose our time :)

**Comment by dpo** [comment]
> I am quite unhappy of all the checks in homebrew-science that sometimes add more work for us, and we even end up disabling some test, after having reported the errors to an unresponsive upstream ... there are better ways to loose our time :)

In my experience, it's often the tool's own unit tests that fail after an upgrade, not what's in the formula's `test` method (or if it is, it's easily fixed). I do see value in filtering tools that have failing unit tests. If the tool's own unit tests don't pass, I feel we shouldn't ship that formula until upstream fixed it. If upstream is unresponsive, there's no reason for us to ship their tool.

Still in my experience, many users don't check correctness for themselves and simply open issues here (as opposed to the upstream tracker) when there's a problem. Octave developers in particular have been extremely proactive in helping resolve issues here with both octave itself and the homebrew formula.

**Comment by fxcoudert** [comment]
sundials migrated to core
@BrewTestBot test this please

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thanks @schoeps for the pull request!

**Comment by romanbsd** [comment]
running `octave` after `brew upgrade` requires java 6 on Sierra (I have JDK 8 installed). Is it intentional? Rebuilding it locally without java...

**Comment by ilovezfs** [comment]
@romanbsd please open an issue and fill out the template completely.

**Pull request closed without merging**